### ISSUE TITLE
Change setup to run php files instead of html

### ISSUE THREAD

**Issue by adamgilmour** [ryanbenson/Harvest#25]
First of all, love the boilerplate - 10 times faster than the one I'm currently using. Great work!

I use a lot of @includes in my templates for headers/footer an such. For that I'd need the index file to be .php format. Just wondering what's the best way to go about altering the gulpfile for this and for browsersync to serve these files to the localhost?

Or...is there a better way to include html files, in html files that I don't know about?

Thanks :)

**Comment by ryanbenson** [comment]
I've actually done this before! It's not that hard :)

Go here: https://github.com/ryanbenson/Harvest/blob/master/gulpfile.js#L206 and change the 'app/*.html' to 'app/*.php'

Then go here: https://github.com/ryanbenson/Harvest/blob/master/gulpfile.js#L141 and change the `html` task to the same thing. From 'app/*.html' to 'app/*.php'

That will change all of the watchers to watch PHP files instead of HTML ones, and still retain the browsersync feature :) Let me know if that works or not!

Also, that won't watch PHP files in sub directories. If you have something like an in includes or classes folder, you can either add those individually to be watched or watch for PHP files recursively. Some of the other Gulp tasks do that if you need a reference :)

**Comment by adamgilmour** [comment]
Ah yes, I think I tried the same thing but the issue I was having was with the browser serving the php file, which would return with `Cannot GET /`
I decided to look down another templating avenue in the form of [Nunjucks](https://mozilla.github.io/nunjucks/) which is even better for what I need it for in my projects. I am now however having a problem with browsersync refreshing too quickly, before the Nunjuck files have compiled, warranting a double save for html changes to show. 

If the build has a few templates/pages it seem to refresh the content fine once saved but if more templates/pages files are added, the nunjucks task can't compile in time before browsersync refreshes. I've tried changing the browsersync reloadDelay option which doesn't to have any effect on the reload speed.

The gulpfile is below if you've got any ideas? (btw, edited for personal preferences):

```
var app, base, concat, directory, gulp, gutil, hostname, path, refresh, sass, uglify, imagemin, minifyCSS, del, browserSync, autoprefixer, gulpSequence, shell, sourceMaps, plumber, nunjucksRender;

var autoPrefixBrowserList = ['last 2 version', 'safari 5', 'ie 8', 'ie 9', 'opera 12.1', 'ios 6', 'android 4'];

gulp        = require('gulp');
gutil       = require('gulp-util');
concat      = require('gulp-concat');
uglify      = require('gulp-uglify');
sass        = require('gulp-sass');
sourceMaps  = require('gulp-sourcemaps');
imagemin    = require('gulp-imagemin');
minifyCSS   = require('gulp-minify-css');
browserSync = require('browser-sync');
autoprefixer = require('gulp-autoprefixer');
gulpSequence = require('gulp-sequence').use(gulp);
shell       = require('gulp-shell');
plumber     = require('gulp-plumber');
nunjucksRender = require('gulp-nunjucks-render');

gulp.task('browserSync', function() {
    browserSync({
        server: {
            baseDir: "app/"
        },
        options: {
            reloadDelay: 250
        },
        notify: false
    });
});

gulp.task('images', function(tmp) {
    gulp.src(['app/images/*.jpg', 'app/images/*.png'])
        .pipe(plumber())
        .pipe(imagemin({ optimizationLevel: 5, progressive: true, interlaced: true }))
        .pipe(gulp.dest('app/images'));
});

gulp.task('images-deploy', function() {
    gulp.src('app/images/**/*')
        .pipe(plumber())
        .pipe(gulp.dest('dist/images'));
});

gulp.task('js', function() {
    return gulp.src(['app/js/src/vendor/**/*.js','app/js/src/nublue/**/*.js', 'app/js/src/**/*.js'])
                .pipe(plumber())
                .pipe(concat('app.js'))
                .on('error', gutil.log)
                .pipe(gulp.dest('app/scripts'))
                .pipe(browserSync.reload({stream: true}));
});

gulp.task('js-deploy', function() {
    return gulp.src(['app/js/src/vendor/**/*.js', 'app/js/src/**/*.js'])
                .pipe(plumber())
                .pipe(concat('app.js'))
                .pipe(uglify())
                .pipe(gulp.dest('dist/js'));
});

gulp.task('css', function() {
    return gulp.src('app/css/scss/init.scss')
                .pipe(plumber({
                  errorHandler: function (err) {
                    console.log(err);
                    this.emit('end');
                  }
                }))
                .pipe(sourceMaps.init())
                .pipe(sass({
                      errLogToConsole: true,
                      includePaths: [
                          'app/css/scss/'
                      ]
                }))
                .pipe(autoprefixer({
                   browsers: autoPrefixBrowserList,
                   cascade:  true
                }))
                .on('error', gutil.log)
                .pipe(concat('styles.css'))
                .pipe(sourceMaps.write())
                .pipe(gulp.dest('app/css'))
                .pipe(browserSync.reload({stream: true}));
});

gulp.task('css-deploy', function() {
    return gulp.src('app/css/scss/init.scss')
                .pipe(plumber())
                .pipe(sass({
                      includePaths: [
                          'app/css/scss',
                      ]
                }))
                .pipe(autoprefixer({
                  browsers: autoPrefixBrowserList,
                  cascade:  true
                }))
                .pipe(concat('styles.css'))
                .pipe(minifyCSS())
                .pipe(gulp.dest('dist/css'));
});

gulp.task('nunjucks', function() {
  return gulp.src('app/pages/**/*.html')
    .pipe(nunjucksRender({
        path: ['app/templates']
      }))
    .pipe(gulp.dest('app'))
    .pipe(plumber())
    .pipe(browserSync.reload({stream: true}))
});

gulp.task('html-deploy', function() {
    gulp.src(['app/*', '!app/{pages,pages/**}', '!app/{templates,templates/**}'])
        .pipe(plumber())
        .pipe(gulp.dest('dist'));

    gulp.src('app/fonts/**/*')
        .pipe(plumber())
        .pipe(gulp.dest('dist/fonts'));

    gulp.src(['app/css/*.css', '!app/css/styles.css'])
        .pipe(plumber())
        .pipe(gulp.dest('dist/css'));
});

gulp.task('clean', function() {
    return shell.task([
      'rm -rf dist'
    ]);
});

gulp.task('scaffold', function() {
  return shell.task([
      'mkdir dist',
      'mkdir dist/fonts',
      'mkdir dist/images',
      'mkdir dist/js',
      'mkdir dist/css'
    ]
  );
});

gulp.task('default', ['browserSync', 'js', 'css', 'nunjucks'], function() {
    gulp.watch('app/js/src/**', ['js']);
    gulp.watch('app/css/scss/**', ['css']);
    gulp.watch('app/images/**', ['images']);
    gulp.watch('app/**/*.html', ['nunjucks']);
});

gulp.task('deploy', gulpSequence('clean', 'scaffold', ['js-deploy', 'css-deploy', 'images-deploy'], 'html-deploy'));
```

Thanks for your help :)

**Comment by ryanbenson** [comment]
I forgot about this lol, my apologies.

Are you running something like MAMP/WAMP as well?

**Comment by adamgilmour** [comment]
Haha, no worries. Running xammp at the moment but have used wamp in the past

**Comment by ryanbenson** [comment]
I couldn't replicate your setup (with Nunchucks), but I was able to use the vanilla version and get it working with MAMP and BrowserSync. I had to make a few minor edits to get it to work.

* I have my MAMP set to the app directory, not root directory
* I changed the browserSync setup on line #16 
* I changed BrowserSync to use proxy instead of a server
* I changed the watcher on the `default` task to use a different event
* Minor edits to the tasks to watch for PHP files

Once BrowserSync opens a new tab, I opened a `test.php` file and started changing the contents and it reloaded as expected. I hope it helps! Let me know if you can't get it to work. I was using this StackOverflow article as a reference, which might be useful too:

http://stackoverflow.com/questions/39657604/using-gulps-browser-sync-with-mamp-in-localhost-testing-environment

Here's are the changes, and I'll dump the full file too.

Changes:
```
browserSync = require('browser-sync').create();

gulp.task('browserSync', function() {
    browserSync.init({
        proxy: "http://localhost/",
        options: {
            reloadDelay: 250
        },
        notify: false
    });
});

gulp.task('html', function() {
    //watch any and all HTML files and refresh when something changes
    return gulp.src('app/*.php')
        .pipe(plumber())
        .pipe(browserSync.reload({stream: true}))
        //catch errors
        .on('error', gutil.log);
});

gulp.task('default', ['browserSync', 'scripts', 'styles'], function() {
    //a list of watchers, so it will watch all of the following files waiting for changes
    gulp.watch('app/scripts/src/**', ['scripts']);
    gulp.watch('app/styles/scss/**', ['styles']);
    gulp.watch('app/images/**', ['images']);
    // gulp.watch('app/*.php', ['html']);
    gulp.watch("app/*.php").on("change", browserSync.reload);
});
```


Full Gulpfile:
```
//initialize all of our variables
var app, base, concat, directory, gulp, gutil, hostname, path, refresh, sass, uglify, imagemin, minifyCSS, del, browserSync, autoprefixer, gulpSequence, shell, sourceMaps, plumber;

var autoPrefixBrowserList = ['last 2 version', 'safari 5', 'ie 8', 'ie 9', 'opera 12.1', 'ios 6', 'android 4'];

//load all of our dependencies
//add more here if you want to include more libraries
gulp        = require('gulp');
gutil       = require('gulp-util');
concat      = require('gulp-concat');
uglify      = require('gulp-uglify');
sass        = require('gulp-sass');
sourceMaps  = require('gulp-sourcemaps');
imagemin    = require('gulp-imagemin');
minifyCSS   = require('gulp-minify-css');
browserSync = require('browser-sync').create();
autoprefixer = require('gulp-autoprefixer');
gulpSequence = require('gulp-sequence').use(gulp);
shell       = require('gulp-shell');
plumber     = require('gulp-plumber');

gulp.task('browserSync', function() {
    browserSync.init({
        proxy: "http://localhost/",
        options: {
            reloadDelay: 250
        },
        notify: false
    });
});


//compressing images & handle SVG files
gulp.task('images', function(tmp) {
    gulp.src(['app/images/*.jpg', 'app/images/*.png'])
        //prevent pipe breaking caused by errors from gulp plugins
        .pipe(plumber())
        .pipe(imagemin({ optimizationLevel: 5, progressive: true, interlaced: true }))
        .pipe(gulp.dest('app/images'));
});

//compressing images & handle SVG files
gulp.task('images-deploy', function() {
    gulp.src(['app/images/**/*', '!app/images/README'])
        //prevent pipe breaking caused by errors from gulp plugins
        .pipe(plumber())
        .pipe(gulp.dest('dist/images'));
});

//compiling our Javascripts
gulp.task('scripts', function() {
    //this is where our dev JS scripts are
    return gulp.src(['app/scripts/src/_includes/**/*.js', 'app/scripts/src/**/*.js'])
                //prevent pipe breaking caused by errors from gulp plugins
                .pipe(plumber())
                //this is the filename of the compressed version of our JS
                .pipe(concat('app.js'))
                //catch errors
                .on('error', gutil.log)
                //where we will store our finalized, compressed script
                .pipe(gulp.dest('app/scripts'))
                //notify browserSync to refresh
                .pipe(browserSync.reload({stream: true}));
});

//compiling our Javascripts for deployment
gulp.task('scripts-deploy', function() {
    //this is where our dev JS scripts are
    return gulp.src(['app/scripts/src/_includes/**/*.js', 'app/scripts/src/**/*.js'])
                //prevent pipe breaking caused by errors from gulp plugins
                .pipe(plumber())
                //this is the filename of the compressed version of our JS
                .pipe(concat('app.js'))
                //compress :D
                .pipe(uglify())
                //where we will store our finalized, compressed script
                .pipe(gulp.dest('dist/scripts'));
});

//compiling our SCSS files
gulp.task('styles', function() {
    //the initializer / master SCSS file, which will just be a file that imports everything
    return gulp.src('app/styles/scss/init.scss')
                //prevent pipe breaking caused by errors from gulp plugins
                .pipe(plumber({
                  errorHandler: function (err) {
                    console.log(err);
                    this.emit('end');
                  }
                }))
                //get sourceMaps ready
                .pipe(sourceMaps.init())
                //include SCSS and list every "include" folder
                .pipe(sass({
                      errLogToConsole: true,
                      includePaths: [
                          'app/styles/scss/'
                      ]
                }))
                .pipe(autoprefixer({
                   browsers: autoPrefixBrowserList,
                   cascade:  true
                }))
                //catch errors
                .on('error', gutil.log)
                //the final filename of our combined css file
                .pipe(concat('styles.css'))
                //get our sources via sourceMaps
                .pipe(sourceMaps.write())
                //where to save our final, compressed css file
                .pipe(gulp.dest('app/styles'))
                //notify browserSync to refresh
                .pipe(browserSync.reload({stream: true}));
});

//compiling our SCSS files for deployment
gulp.task('styles-deploy', function() {
    //the initializer / master SCSS file, which will just be a file that imports everything
    return gulp.src('app/styles/scss/init.scss')
                .pipe(plumber())
                //include SCSS includes folder
                .pipe(sass({
                      includePaths: [
                          'app/styles/scss',
                      ]
                }))
                .pipe(autoprefixer({
                  browsers: autoPrefixBrowserList,
                  cascade:  true
                }))
                //the final filename of our combined css file
                .pipe(concat('styles.css'))
                .pipe(minifyCSS())
                //where to save our final, compressed css file
                .pipe(gulp.dest('dist/styles'));
});

//basically just keeping an eye on all HTML files
gulp.task('html', function() {
    //watch any and all HTML files and refresh when something changes
    return gulp.src('app/*.php')
        .pipe(plumber())
        .pipe(browserSync.reload({stream: true}))
        //catch errors
        .on('error', gutil.log);
});

//migrating over all HTML files for deployment
gulp.task('html-deploy', function() {
    //grab everything, which should include htaccess, robots, etc
    gulp.src('app/*')
        //prevent pipe breaking caused by errors from gulp plugins
        .pipe(plumber())
        .pipe(gulp.dest('dist'));

    //grab any hidden files too
    gulp.src('app/.*')
        //prevent pipe breaking caused by errors from gulp plugins
        .pipe(plumber())
        .pipe(gulp.dest('dist'));

    gulp.src('app/fonts/**/*')
        //prevent pipe breaking caused by errors from gulp plugins
        .pipe(plumber())
        .pipe(gulp.dest('dist/fonts'));

    //grab all of the styles
    gulp.src(['app/styles/*.css', '!app/styles/styles.css'])
        //prevent pipe breaking caused by errors from gulp plugins
        .pipe(plumber())
        .pipe(gulp.dest('dist/styles'));
});

//cleans our dist directory in case things got deleted
gulp.task('clean', function() {
    return shell.task([
      'rm -rf dist'
    ]);
});

//create folders using shell
gulp.task('scaffold', function() {
  return shell.task([
      'mkdir dist',
      'mkdir dist/fonts',
      'mkdir dist/images',
      'mkdir dist/scripts',
      'mkdir dist/styles'
    ]
  );
});

//this is our master task when you run `gulp` in CLI / Terminal
//this is the main watcher to use when in active development
//  this will:
//  startup the web server,
//  start up browserSync
//  compress all scripts and SCSS files
gulp.task('default', ['browserSync', 'scripts', 'styles'], function() {
    //a list of watchers, so it will watch all of the following files waiting for changes
    gulp.watch('app/scripts/src/**', ['scripts']);
    gulp.watch('app/styles/scss/**', ['styles']);
    gulp.watch('app/images/**', ['images']);
    // gulp.watch('app/*.php', ['html']);
    gulp.watch("app/*.php").on("change", browserSync.reload);
});

//this is our deployment task, it will set everything for deployment-ready files
gulp.task('deploy', gulpSequence('clean', 'scaffold', ['scripts-deploy', 'styles-deploy', 'images-deploy'], 'html-deploy'));
```

**Comment by adamgilmour** [comment]
Just had the chance to properly test this and it works great for a php file setup. Good work 👍 

Although, since finding the Nunjucks templating system I've decided to switch over to html templating, removing a need for php files. I've found this method a lot quicker and more versatile when importing other templates. For example - a 'layout.njk' template which contains the global wrapper (html, head, body etc) can be used, not possible with a php include templating method.

At the moment this html templating method works great but as I've mentioned in the previous comment, browsersync refreshes too quickly or doesn't wait for the task to finish compiling. I'll keep trying to fix it though!

I've also added a svg sprite feature to the mix which takes all svg files from a folder and adds it to a svg symbol based sprite, stripping path/styles etc. I'll be doing a fork when I've got time.

Cheers for your help 👍
